### PR TITLE
Refactor and parallelize MaxText test runner

### PR DESCRIPTION
**Pull request by shralex**
This change moves train, decode and integration tests into separate files, so they can be tested using pytest rather than explicitly being invoked from UnitTests.yml, and splits up various tasks in UnitTests.yml into separate jobs (with explicit dependencies) that execute concurrently. As a result, the run time of tests in MaxText PRs is reduced from ~40min to ~24min.

Before:

image

After:

Screenshot 2024-12-28 at 7 52 59 PM

The following changes are made:
1. All explicit test invocations are moved out of UnitTests.yml. The file is renamed to RunTests.yml.
3. end-to-end tests are placed in a new folder tests/integration_tests
4. RunTests.yml is refactored to use shared modules in 2 new files (one for building/uploading images and another for running tests) and explicit jobs are created for various types of tests. Previously, building a gpu image blocked TPU tests, and all TPU tests ran sequentially. Now GPU image only blocks GPU tests, and unit / integration tests run concurrently. 
5.  Pytest marker tpu is renamed to tpu_only without change in meaning, and we also add a gpu_only marker and an integration_test marker.
6. The frequency of unit-test runs from every 2 hours to every 6 hours, which seems sufficient. 
7. This PR fixes a small bug where we were invoking end_to_end/test_generate_param_only_checkpoint.sh twice with  effectively identical arguments.

There are multiple opportunities for further speed-up, not done in this PR.  For example, utilizing pytest parallelization support, reducing the time to build a GPU image (which is currently ~13 min), avoiding downloading images for every test stage, etc.

FIXES: b/385505333
FIXES: b/385510862

# Tests

This PR refactors and introduces new unit tests.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: .github/workflows/UnitTests.yml [hunk 1]
```diff
@@ -91,87 +89,19 @@ jobs:
...
-      run: python3 MaxText/train.py MaxText/configs/base.yml run_name=runner_$(date +%Y-%m-%d-%H-%M-%S) base_output_directory=gs://runner-maxtext-logs dataset_path=gs://maxtext-dataset quantization=int8 steps=2 enable_checkpointing=false attention=${{ matrix.device.attention }}
-    - name: Test fp8_training
-      run: python3 MaxText/train.py MaxText/configs/base.yml run_name=runner_$(date +%Y-%m-%d-%H-%M-%S) base_output_directory=gs://runner-maxtext-logs dataset_path=gs://maxtext-dataset quantization=fp8 steps=2 enable_checkpointing=false attention=${{ matrix.device.attention }}
-    - name: Test train.py with dropout
-      run: python3 MaxText/train.py MaxText/configs/base.yml run_name=runner_$(date +%Y-%m-%d-%H-%M-%S) base_output_directory=gs://runner-maxtext-logs dataset_path=gs://maxtext-dataset steps=2 enable_checkpointing=false attention=${{ matrix.device.attention }} max_target_length=128 per_device_batch_size=1 dropout_rate=0.02
-    - name: Test generate_param_only_checkpoint
-      run: bash end_to_end/test_generate_param_only_checkpoint.sh -r runner_$(date +%Y-%m-%d-%H-%M-%S) -o gs://runner-maxtext-logs -d gs://maxtext-dataset -i 4 -a ${{ matrix.device.attention }}
```
Comment (shralex) [inline .github/workflows/UnitTests.yml:117]: Note that this is equivalent to the following command, since int8 is the default quantization in 
 test_generate_param_only_checkpoint.sh

File: .github/workflows/UnitTests.yml [hunk 2]
```diff
@@ -23,8 +23,8 @@ on:
     branches: [ "main" ]
   workflow_dispatch:
   schedule:
-    # Run the job every 2 hours
-    - cron:  '0 */2 * * *'
+    # Run the job every 6 hours
```
[the remark this answers is not in the document]
Comment (shralex) [reply .github/workflows/UnitTests.yml]: IMHO given that this also runs on every PR, 6 hours is plenty. For reference, MaxDiffusion has it set to every 12 hours, while Jax only runs its CI once or twice daily, depending on the workflow. I changed it to 4 hours as a compromise for now :) but please let me know if you still disagree.

File: MaxText/tests/decode_int8_test.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+"""Short test for decode with int8 quantization"""
+import os
+import unittest
+import pytest
+from decode import main as decode_main
+from absl.testing import absltest
+
+
+class Train(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/decode_int8_test.py:25]: good catch, thanks!

File: MaxText/tests/decode_pdb_lt_1_test.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+"""Short test for decode with per_device_batch_size < 1"""
+import os
+import unittest
+import pytest
+from decode import main as decode_main
+from absl.testing import absltest
+
+
+class Train(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/decode_pdb_lt_1_test.py:25]: Done

File: MaxText/tests/decode_test.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+"""Short test for decode"""
+import os
+import unittest
+import pytest
+from decode import main as decode_main
+from absl.testing import absltest
+
+
+class Train(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/decode_test.py:25]: Done

File: MaxText/tests/train_pdb_lt_1_test.py
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/train_pdb_lt_1_test.py:1]: I moved all the new train configs into train_tests.py and all the new decode configs into decode_tests.py.

File: MaxText/tests/train_int8_test.py
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/train_int8_test.py:1]: I moved all the new train configs into train_tests.py and all the new decode configs into decode_tests.py.

File: MaxText/tests/integration_tests/shmap_collective_matmul_test.py
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/tests/integration_tests/shmap_collective_matmul_test.py:1]: Done

File: MaxText/tests/train_tests.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+          "steps=2",
+          "enable_checkpointing=False",
+          "dataset_type=hf",
+          "hf_path=parquet",
+          r"hf_train_files=gs://maxtext-dataset/hf/c4/c4-train-00000-of-01637.parquet",
+          r"tokenizer_path=google-t5/t5-large",
+      ]
+  }
+
+  @pytest.mark.tpu_only
```
Comment (gobbleturk) [inline MaxText/tests/train_tests.py:111]: What do you think of running separate @pytest for each config, e.g. something like 

AOT tests: https://github.com/AI-Hypercomputer/maxtext/blob/4699f4f2da2a3ff4956e89de8fb561550ecbbb16/MaxText/tests/train_compile_test.py#L27-L80

I prefer this since:
* Each test is more self contained, don't need to find the right config
* I think if a test fails in the current "for loop" implementation we won't get a clear error mesage as to which test really failed - we will get "test_tpu_config" failed and we will need to carefully scan the logs for the "Running TPU test for config...." message

File: MaxText/tests/gpt3_test.py [hunk 7]
```diff
@@ -85,7 +85,7 @@ def setUp(self):
     }
     self.model_vars = init_random_model_vars(self.model, self.rng, self.example_batch)
 
-  @pytest.mark.tpu
+  @pytest.mark.tpu_only
   def test_logits_numerically(self):
```
[the remark this answers is not in the document]
Comment (ZhiyuLi-goog) [reply MaxText/tests/gpt3_test.py:91]: Yeap. You are right. This test is simply to avoid numerical discrepancy caused by any unexpected implementation changes in sharing module like attention or linear layers.

We don't care about the exact value of the model weight. A fixed random seed ensures a frozen model weights. And we do expect the same results with the same model weights as long as the implementation is not changed.  

The root cause might be relevant to a changing of random state, which might be affected by other test in parallel. I propose to double check random state in parallel tests first. Otherwise, we can switch to ckpt loading if there's hard conflict between random state and parallel testing.

<br/><br/>
> In an offline process you pass the same random model weights to get ground truth in a paxml implementation (how is this done, how do you ensure the same model weights are used?)

This was completed offline during the #293 see [numerical test](https://colab.sandbox.google.com/drive/1PuOiJTd9dGXheWzNpSYU_MEwpN-i6dyT?usp=sharing). Paxml environment is hard to setup and might introduce some conflicts.

<br/><br/>
> Won't this test fail if we pass different rng or jax changes its rng implementation since the model weights will change? 

Yeap. The goal is to verify the same logit outputs with fixed rng and  random weights while we may not care too much about the exact value of random weights. I think rng implementation might probably pretty stable in jax. 

<br/><br/>
> I think this correctness test makes more sense as an e2e test that loads a checkpoint from GCS.

You are right. It would be ideal to avoid dependency to random initialization. However, it might be hard to maintain for a long time since we need to change orbax ckpt each time
* we change the layer name
* change ckpt implementation
* some updates in orbax library

This happened in last mlperf 4.1 submission.

As for e2e test, there's a functional [e2e test](https://github.com/AI-Hypercomputer/maxtext/blob/main/end_to_end/tpu/test_gpt3.sh) for ckpt conversion and loading.

File: MaxText/tests/train_tests.py (hunk 6 above)
Comment (shralex) [reply MaxText/tests/train_tests.py:111]: Done.

**Review by gobbleturk — APPROVED** [review]
